scarlet2/nn.py

```diff
@@ -0,0 +1,156 @@
+    if trained_model in ('hsc', 'None'):
+        if data_size <= 32:
+            pad_gap = 32 - data_size
+            ScoreNet = HSC_ScoreNet32
+        else:
+            pad_gap = 64 - data_size
+            ScoreNet = HSC_ScoreNet64
+
+    
+    # select the ZTF trained model
+    elif trained_model == 'ztf':
+        if data_size <= 32:
+            pad_gap = 32 - data_size
+            ScoreNet = ZTF_ScoreNet32
+        else:
+            pad_gap = 64 - data_size
+            ScoreNet = ZTF_ScoreNet64
+    # select the custom trained model
+    else:
+        ScoreNet = trained_model
+    # dont pad if we dont need to
+    if pad_gap == 0:
+        pad = False
+    # calculate how much to pad    
+    if pad_gap % 2 == 0:
+        pad_lo = pad_hi = int(pad_gap / 2)
+    else:
+        pad_lo = int(pad_gap // 2)
+        pad_hi = int(pad_lo + 1)
+    # perform the zero-padding
+    if pad:
+        if jnp.ndim(x) == 3:
+            x[0] = jnp.pad(x[0], ((pad_lo, pad_hi), (pad_lo, pad_hi)), 'constant', constant_values=0) 
+        else:
+            x = jnp.pad(x, ((pad_lo, pad_hi), (pad_lo, pad_hi)), 'constant', constant_values=0)
+    return x, ScoreNet, pad_lo, pad_hi , pad
+    
+# reverse pad back to original size
+def pad_back(x, pad_lo, pad_hi):
+    """Removes the zero-padding from the input image"""
+    if jnp.ndim(x) > 2: 
+        x[0] = x[pad_lo:-pad_hi, pad_lo:-pad_hi]
+    else:
+        x = x[pad_lo:-pad_hi, pad_lo:-pad_hi]
+    return x
+
+# calculate score function (jacobian of log-probability)
+def calc_grad(x, trained_model):
+    """Calculates the gradient of the log-probability using the ScoreNet model"""
+    # perform padding if needed
+    t = 0.0 # corresponds to noise free gradient
+    x = jnp.float32(x) # cast to float32
+    x, ScoreNet, pad_lo, pad_hi, pad = pad_fwd(x, trained_model)
+    assert (x.shape[1] % 32) == 0, f"image size must be 32 or 64, got: {x.shape[1]}"
+    if jnp.ndim(x) == 2:
+        x = jnp.expand_dims(x, axis=0)
+        nn_grad = ScoreNet(x, t=t)
+        nn_grad = jnp.squeeze(nn_grad, axis=0)
+    else:
+        nn_grad = ScoreNet(x, t=t)
+    # return to original size
+    if pad: nn_grad = pad_back(nn_grad, pad_lo, pad_hi)
+    return nn_grad
+
+
+# ------------------------------------------------- #
+#   transformation corrections in training space    #
+# ------------------------------------------------- #
+# jax gradient function
+def vgrad(f, x):
+    y, vjp_fn = vjp(f, x)
+    return vjp_fn(jnp.ones(y.shape))[0]
+
+# transform to training space
+def transform(x):
+    sigma_y = 0.10
+    return jnp.log(x + 1) / sigma_y
+# ------------------------------------------------- #
+
+
+# inheritance from Distribution class
+class NNPrior(dist.Distribution):
+    support = constraints.real_vector
+
+    """Prior distribution based on a neural network"""
+
+    # construct custom vector-jacobian product
+
+    def __init__(self, trained_model='None', shape=1, validate_args=None):
```

Comment:
`trained_model` -> `model`, which is the actual model or its file path. And let's store it in `self.model` rather than a global variable. Let's also store `self.shape`, but we should be able to get that from the model itself, right?